src/Microsoft.AspNet.Mvc.TagHelpers/Internal/AttributeMatcher.cs

```diff
@@ -93,20 +93,26 @@ public static class AttributeMatcher
             var presentAttributes = new List<string>();
             var missingAttributes = new List<string>();
 
-            foreach (var attribute in requiredAttributes)
+            foreach (var requiredAttribute in requiredAttributes)
             {
-                if (!context.AllAttributes.ContainsName(attribute) ||
```

Comment:
The code change made something more visible that existed before. Could certainly deal w/ in a follow-up issue. Good thought @Eilon.

Separately I remembered some of the history here: We were originally thinking about required attribute as a subset of bound attributes. The code definitely also pre-dated handling of minimized attributes. Enough has changed since for the checks (now at [line 107](https://github.com/aspnet/Mvc/pull/3637/files#diff-a8ba4397ef9cafc42bb0d852329baa66R107)) to be irrelevant and often-confusing esp. for attributes that are required but not bound.